internal/controlplane/xds.go

```diff
@@ -19,51 +20,48 @@ import (
 	envoy_extensions_access_loggers_grpc_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/access_loggers/grpc/v3"
 	envoy_extensions_transport_sockets_tls_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
 	envoy_service_discovery_v3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
-	"github.com/golang/protobuf/ptypes"
-	"github.com/golang/protobuf/ptypes/any"
 	"golang.org/x/net/nettest"
-	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
 
 	"github.com/pomerium/pomerium/config"
 	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/pkg/cryptutil"
 )
 
-func (srv *Server) buildDiscoveryResponse(version string, typeURL string, options *config.Options) (*envoy_service_discovery_v3.DiscoveryResponse, error) {
-	switch typeURL {
-	case "type.googleapis.com/envoy.config.listener.v3.Listener":
-		listeners := buildListeners(options)
-		anys := make([]*any.Any, len(listeners))
-		for i, listener := range listeners {
-			a, err := ptypes.MarshalAny(listener)
-			if err != nil {
-				return nil, status.Errorf(codes.Internal, "error marshaling type to any: %v", err)
-			}
-			anys[i] = a
+const (
+	clusterTypeURL  = "type.googleapis.com/envoy.config.cluster.v3.Cluster"
+	listenerTypeURL = "type.googleapis.com/envoy.config.listener.v3.Listener"
+)
+
+func (srv *Server) buildDiscoveryResources() map[string][]*envoy_service_discovery_v3.Resource {
+	resources := map[string][]*envoy_service_discovery_v3.Resource{}
+	cfg := srv.currentConfig.Load()
+	for _, cluster := range srv.buildClusters(&cfg.Options) {
+		any, err := anypb.New(cluster)
+		if err != nil {
+			// anypb will only panic due to developer error.
```

Comment:
My point was it was safe to ignore the error because it doesn't happen at runtime. When it errors it means the function was used improperly by the developer. (a condition where I view it as acceptable to panic)

I removed the panic.

Reply:
I understood, I was just joking :)